### PR TITLE
CE-2035 Completed development and testing of UCM-powershell agent for SQL Server

### PR DESCRIPTION
**Pull request by gagan93jtg**

### CODE REVIEW

File: UCM-Powershell/metric_group.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		"type": "ce_counter",
+		"unit": "Compilations"
+	}, {
+		"name": "sql_recompilations",
+		"label": "SQL Recompilations",
+		"type": "ce_counter",
+		"unit": "Recompilations"
+	}],
+	"frequency": "60"
+}
```
Comment (Neelakshi1808) [inline UCM-Powershell/metric_group.json]: Include End of File

File: UCM-Powershell/dashboard.json [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+					"mssql2": [
+						[9, "sql_recompilations"]
+					]
+				}
+			}
+		},
+		"order": ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
+	},
+	"name": "SQL-Server"
+}
```
Comment (Neelakshi1808) [inline UCM-Powershell/dashboard.json]: End of File

File: UCM-Powershell/config.xml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+		</Server1>
+		<Server1>
+			<InstanceName>InstanceNameGoesHerre</InstanceName>
+			<Hostname>HostnameGoesHere</Hostname>
+			<Username>UsernameGoesHereIfSQLAuthentication</Username>
+			<Password>PasswordGoesHereIfSQLAuthentication</Password>
+			<SystemIdentifier>UniqueSystemIdentifierGoesHere</SystemIdentifier>
+		</Server1>
+	</SQLServers>
+</Settings>
```
Comment (Neelakshi1808) [inline UCM-Powershell/config.xml]: EOF

File: UCM-Powershell/config.xml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<Settings>
+	<UptimeCloudMonitor>
+		<Apikey>AbCdEf1234567890</Apikey>
```
Comment (Neelakshi1808) [inline UCM-Powershell/config.xml]: Remove this dummy API key, and put text here : User's API key

File: UCM-Powershell/config.xml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	</SQLServersCommonSettings>
+	<SQLServers>
+		<Server1>
+			<InstanceName>InstanceNameGoesHerre</InstanceName>
+			<Hostname>HostnameGoesHere</Hostname>
+			<Username>UsernameGoesHereIfSQLAuthentication</Username>
+			<Password>PasswordGoesHereIfSQLAuthentication</Password>
+			<SystemIdentifier>UniqueSystemIdentifierGoesHere</SystemIdentifier>
+		</Server1>
+		<Server1>
```
Comment (Neelakshi1808) [inline UCM-Powershell/config.xml]: Remove this second server prototype or name it properly, user might fill the first one and leave this one as it is.

File: UCM-Powershell/config.xml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+		<MetricGroupLabel>SQL Server Metrics</MetricGroupLabel>
+		<Frequency>15</Frequency>
+	</SQLServersCommonSettings>
+	<SQLServers>
+		<Server1>
+			<InstanceName>InstanceNameGoesHerre</InstanceName>
+			<Hostname>HostnameGoesHere</Hostname>
+			<Username>UsernameGoesHereIfSQLAuthentication</Username>
+			<Password>PasswordGoesHereIfSQLAuthentication</Password>
+			<SystemIdentifier>UniqueSystemIdentifierGoesHere</SystemIdentifier>
```
Comment (Neelakshi1808) [inline UCM-Powershell/config.xml]: Text for this field should be more clear indicating this is the name with which things will be monitored in UCM

File: UCM-Powershell/metric_group.json (hunk 1 above)
Comment (gagan93jtg) [inline UCM-Powershell/metric_group.json]: resolved

File: UCM-Powershell/dashboard.json (hunk 2 above)
Comment (gagan93jtg) [inline UCM-Powershell/dashboard.json]: resolved

File: UCM-Powershell/config.xml (hunk 3 above)
Comment (gagan93jtg) [inline UCM-Powershell/config.xml]: resolved

File: UCM-Powershell/config.xml (hunk 4 above)
Comment (gagan93jtg) [inline UCM-Powershell/config.xml]: resolved

File: UCM-Powershell/config.xml (hunk 5 above)
Comment (gagan93jtg) [inline UCM-Powershell/config.xml]: second block removed. Moreover, file was renamed as config-sample.xml. When installation script runs, a proper config.xml will be generated based on user inputs.

File: UCM-Powershell/config.xml (hunk 6 above)
Comment (gagan93jtg) [inline UCM-Powershell/config.xml]: resolved

File: UCM-Powershell/UCM-SQL-Monitor.ps1 [hunk 7]
```diff
@@ -61,10 +64,23 @@ ForEach ($server in $SQLServers.ChildNodes)
...
+  $StopJob = Test-Path -Path "$env:temp\stop-ucm-monitor.txt"
+  if ($StopJob)
+  {
+    Write-Log "Stopping job as requested by user."
+    Remove-Item "$env:temp\stop-ucm-monitor.txt"
+    exit 0
+  }
+  if ($SleepTime % 60 -eq 0)
+  {
+    Write-Host "Running $($jobs.count) workers. Check copperegg-metrics.log for any worker logs."
```
Comment (Neelakshi1808) [inline UCM-Powershell/UCM-SQL-Monitor.ps1]: What is we are starting script and the file is present ? The script will kill itself.

File: UCM-Powershell/UCM-SQL-Monitor.ps1 [hunk 8]
```diff
@@ -61,10 +64,23 @@ ForEach ($server in $SQLServers.ChildNodes)
...
   Write-Log "Spawning worker thread : $($job|out-string)"
 }
+$SleepTime = 0
 while($TRUE)
 {
   $jobs = Get-job
-  Write-Host "Running $($jobs.count) workers. Check copperegg-metrics.log for any worker logs.
-              Press Ctrl+C to stop monitoring"
-  Start-Sleep -s 60
+  $StopJob = Test-Path -Path "$env:temp\stop-ucm-monitor.txt"
```
Comment (Neelakshi1808) [inline UCM-Powershell/UCM-SQL-Monitor.ps1]: What is we are starting script and the file is present ? The script will kill itself.

File: UCM-Powershell-MSSQL/Monitor-Worker.ps1 [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+                     counter_name = 'sql re-compilations/sec' OR
+                     counter_name = 'transaction delay' OR 
+                     counter_name = 'transaction ownership waits' OR 
+                     counter_name = 'transactions' OR 
+                     counter_name = 'write transactions/sec'";
+
+$Request = New-Object System.Net.WebClient
+
+# LogFile path is hardcoded here because worker thread cannot read $PSScriptRoot Variable 
+$LogFile = "$env:programfiles\UCM-Powershell\ucm-metrics.log"
```
Comment (marutinandanpandya) [inline UCM-Powershell-MSSQL/Monitor-Worker.ps1]: remove hard coding of paths

**Pull request merged**